Do not animate hidden `WebView`s

mrobinson opened the pull request:
When a `WebView` is hidden  it should not process animations. This
change makes it so that this does not happen by:

- Tracking visibility in the Constellation. Hidden WebViews are
  throttled and hiding and showing WebViews does not force Pipelines
  into a non-throttled state.
- Tracking visibility in the embedder-side `WebView` data structure.
  Hidden `WebView`s do not get new frame notifications and do not
  trigger animation ticks in the `RefreshDriver`.

Testing: This is a performance optimization so it is difficult to test this via
automated testing.

mrobinson commented:
> * `RunningAppState::notify_focus_changed` should not hide the unfocused webviews, considering the possibility that multiple webview instances are simultaneously visible in the foreground. It might be better to leave `hide` operation for embedders.

Great idea. I was considering doing this as well to fix the newly failing tests (they use popups). It seems we will need some kind of special logic in servoshell to handle that.

coding-joedow commented:
I have another question regarding the throttled state tracked in constellation‘s `BrowsingContext`. The `throttled` state can be transferred from parent pipeline to child pipelines, opener to openee, between history traversal or navigation. However, I can not find where the `throttled` state of `BrowsingContext` is set. Do I miss anything?

Besides, I can not understand the relationship between `BrowsingContext::throttled`, `Pipeline::has_active_document` and `Webview::visible`

minghuaw commented:
Not sure if this is worth investigating or not, but the rayon threadpool used for layout (styling, box tree construction and fragment tree construction) uses `Script(1,1)` thread as one of its worker thread. Given that background webview should be throttled, is it worth revisiting that choice (stylo rayon pool using Script(1,1) as one of the worker thread)?

mrobinson committed:
Do not animate hidden `WebView`s

When a `WebView` is hidden it should not process animations. This change
makes it so that this does not happen by:

- Tracking visibility in the Constellation. Hidden WebViews are
  throttled. Pipelines are always throttled / unthrottled based on
  the hidden state of their containing WebView.
- Tracking visibility in the embedder-side `WebView` data structure.
  Hidden `WebView`s do not get new frame notifications and do not
  trigger animation ticks in the `RefreshDriver`.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson committed:
constellation: Remove `BrowsingContext::throttled`

This is no longer necessary now that `WebView`s track visibility.

Signed-off-by: Martin Robinson <mrobinson@igalia.com>

mrobinson committed:
servoshell: Have `window.open()` in servoshell in new windows

Signed-off-by: Martin Robinson <mrobinson@igalia.com>
